File: bridges/nvda/addon/globalPlugins/nvdaMcpBridge/domain/controllers/commands/type_text.py
# nvdaMcpBridge domain -- TypeTextHandler: insert literal text at the focus.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
# ROLE: command handler for `typeText`, reporting the speech that arrived in the grace window.

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .... import protocol
from .command_handler import CommandHandler
from .observation import speech_entries, state_snapshot

if TYPE_CHECKING:
	from .session_context import SessionContext


class TypeTextHandler(CommandHandler):
	mutates_reader = True

	def execute(self, ctx: SessionContext, request: protocol.Request) -> Any:
		params = protocol.from_dict(protocol.TypeParams, request.params)
		grace = max(0, params.graceMs) / 1000.0
		buffer = ctx.speech_buffer

		if params.announce.strip():
			ctx.announce_to_human(params.announce)

		start_index = buffer.next_index()
		# Logged before injection so a failed attempt is still recorded; only the length, never the text.
		ctx.transcript.typed(len(params.text))
		ctx.adapter_set.text_typer.type_text(params.text)
		buffer.collect_since(start_index, grace)

		entries, from_index, to_index = buffer.entries_since(start_index)
		return protocol.TypeResult(
			# Never the text: typing is how a secret is entered.
			typed=len(params.text),
			speech=speech_entries(entries),
			speechFrom=from_index,
			speechTo=to_index,
			state=state_snapshot(ctx),
		)

File: bridges/voiceover/Sources/VoiceOverBridgeDomain/Controllers/Commands/TypeText.swift
// ROLE: controller for `typeText`: inserts literal text where the focus is, then reports what the
// reader said.
// BUILT BY: Registry. DRIVES: the PermissionBroker and TextTyper ports and the session's
// SpeechBuffer.
// The grant is checked before the injection, which cannot fail visibly; see TextTyper.

import Foundation
import ScreenReaderWire

public final class TypeTextHandler: CommandHandler {
	public let mutatesReader = true

	public init() {}

	public func execute(_ context: SessionContext, _ request: Request) throws -> any Encodable {
		let params = try request.params(as: TypeParams.self)
		let adapters = try readerEdge(context)
		let buffer = try context.speechBuffer()
		let grace = Double(max(0, params.graceMs)) / 1000.0

		// Warned before the grant is asked for and before anything is typed.
		try HumanWarning.honour(context, params.announce)

		try AccessibilityGrant.ensure(adapters.permissions, orElse: "nothing was typed")

		let startIndex = buffer.nextIndex()
		// Recorded before the injection, so a failed attempt is still in the record: the length, never
		// the text.
		context.transcript.typed(Self.count(params.text))
		do {
			try adapters.textTyper.type(params.text)
		} catch let failure as TypingError {
			throw CommandError("the text could not be typed: \(failure.description)")
		}
		let read = buffer.collectSince(startIndex, grace: grace)

		return TypeResult(
			// The length, never the text: this is how a secret is entered.
			typed: Self.count(params.text),
			speech: Observation.speechEntries(read.entries),
			speechFrom: read.fromIndex,
			speechTo: read.toIndex,
			// Nil: this bridge announces no `state` capability.
			state: nil
		)
	}

	/// Unicode scalars, not grapheme clusters: `typed` must match the code-point count the other wire
	/// bindings report.
	static func count(_ text: String) -> Int {
		text.unicodeScalars.count
	}

	private func readerEdge(_ context: SessionContext) throws -> AdapterSet {
		guard let adapters = context.adapters else {
			throw CommandError("text was typed before `hello` built the reader edge")
		}
		return adapters
	}
}
